### PR TITLE
Syntax Highlighting Improvements

### PR DESCRIPTION
**Pull request by hanniavalera**
### This changes visible behavior

This pull request significantly improves CMake syntax highlighting, addressing several long-standing issues and expanding support for modern CMake features. The main focus is on more accurate keyword, variable, and generator expression recognition, as well as eliminating false positives for deprecated keywords. Additionally, a comprehensive new syntax test file has been added to verify these enhancements.

**Syntax Highlighting Improvements:**

* Improved CMake syntax highlighting with command-scoped keyword rules, expanded variable recognition, and better generator-expression support. This also removes obsolete deprecated-keyword entries that previously caused false positives on user-defined names.

**Testing Enhancements:**

* Added a new `test/visual-syntax-test/CMakeLists.txt` file containing a wide range of CMake constructs and edge cases, ensuring that the updated syntax highlighting covers project keywords, variables, boolean values, generator expressions, and module commands, while avoiding incorrect highlighting of user-defined names.

Here are some images of what the highlighting looks now with these changes:
Screenshot 2026-02-18 141715
Screenshot 2026-02-18 141726
Screenshot 2026-02-18 141740
Screenshot 2026-02-19 103024
Screenshot 2026-02-18 141836
Screenshot 2026-02-18 141852

### CODE REVIEW

File: syntaxes/CMake.tmLanguage.json [hunk 1]
```diff
@@ -16,40 +41,394 @@
...
+        { "include": "#cacheVariableReference" }
+      ]
+    },
+    "bracketComment": {
+      "name": "comment.line.number-sign.bracket.cmake",
+      "begin": "#\\[(=*)\\[",
+      "beginCaptures": { "0": { "name": "punctuation.definition.comment.begin.cmake" } },
+      "end": "\\](=*)\\]",
+      "endCaptures": { "0": { "name": "punctuation.definition.comment.end.cmake" } }
+    },
```
[the remark this answers is not in the document]
Comment (hanniavalera) [reply syntaxes/CMake.tmLanguage.json:87]: This is a TextMate limitation as TextMate grammars do not support backreferences in end patterns to reference captures from begin. The \\1 in an end pattern only refers to groups within the end regex itself, not the begin regex. This is a well-known limitation of the TextMate grammar engine (used by VS Code). There is no workaround short of enumerating separate rules for each possible = count (e.g., #[[, #[=[, #[==[, etc.), which is impractical and fragile.

File: syntaxes/CMake.tmLanguage.json [hunk 2]
```diff
@@ -16,40 +41,394 @@
...
+      "match": "#(.*)$",
+      "captures": { "1": { "name": "comment.line.cmake" } }
+    },
+    "bracketArgument": {
+      "name": "variable.parameter.cmake",
+      "begin": "\\[(=*)\\[",
+      "beginCaptures": { "0": { "name": "punctuation.definition.string.begin.cmake" } },
+      "end": "\\](=*)\\]",
+      "endCaptures": { "0": { "name": "punctuation.definition.string.end.cmake" } }
+    },
```
[the remark this answers is not in the document]
Comment (hanniavalera) [reply syntaxes/CMake.tmLanguage.json:99]: Same issue as above.

File: syntaxes/CMake.tmLanguage.json [hunk 3]
```diff
@@ -16,40 +41,394 @@
...
+      "patterns": [
+        { "include": "#nestedParensIf" },
+        { "include": "#commonArgPatterns" },
+        {
+          "comment": "if/elseif operators: comparisons, tests, logic",
+          "name": "keyword.operator.cmake",
+          "match": "\\b(?i:NOT|AND|OR|COMMAND|POLICY|TARGET|TEST|DEFINED|IN_LIST|EXISTS|IS_ABSOLUTE|IS_DIRECTORY|IS_SYMLINK|IS_NEWER_THAN|MATCHES|LESS|GREATER|EQUAL|LESS_EQUAL|GREATER_EQUAL|STRLESS|STRGREATER|STREQUAL|STRLESS_EQUAL|STRGREATER_EQUAL|VERSION_LESS|VERSION_GREATER|VERSION_EQUAL|VERSION_LESS_EQUAL|VERSION_GREATER_EQUAL|PATH_EQUAL)\\b"
+        }
+      ]
+    },
```
[the remark this answers is not in the document]
Comment (hanniavalera) [reply syntaxes/CMake.tmLanguage.json]: addressed, reasonable suggestion

File: syntaxes/CMake.tmLanguage.json [hunk 4]
```diff
@@ -16,40 +41,394 @@
...
+        { "include": "#variableReference" },
+        { "include": "#envVariableReference" },
+        { "include": "#cacheVariableReference" },
+        { "include": "#generatorExpression" }
+      ]
+    },
+    "unquoted": {
+      "name": "string.unquoted.cmake",
+      "begin": "[^\\s()#\"\\\\][^\\s()#\"\\\\]*",
+      "end": "(?=\\s|\\(|\\)|#|\"|\\\\|$)",
```
[the remark this answers is not in the document]
Comment (hanniavalera) [reply syntaxes/CMake.tmLanguage.json:74]: addressed

**Comment by gcampbell-msft** [comment]
Overall this looks good, but it looks like there might still be some edge cases we're not hitting: 

image
Shouldn't the LOG_DOWNLOAD, LOG_UPDATE, etc., be blue as well?

File: syntaxes/CMake.tmLanguage.json [hunk 5]
```diff
@@ -16,40 +41,388 @@
...
+      },
+      "end": "\\)",
+      "endCaptures": { "0": { "name": "punctuation.section.arguments.end.cmake" } },
+      "patterns": [
+        { "include": "#nestedParensIf" },
+        { "include": "#commonArgPatterns" },
+        { "include": "#conditionalOperators" }
+      ]
+    },
+    "scopedForeachCommand": {
```
Comment (gcampbell-msft) [inline syntaxes/CMake.tmLanguage.json]: It feels like the scoped stuff should be combined all into a generic rule, rather than a rule for if-else, foreach, while etc.

**Comment by hanniavalera** [comment]
> Overall this looks good, but it looks like there might still be some edge cases we're not hitting:
> 
> image Shouldn't the LOG_DOWNLOAD, LOG_UPDATE, etc., be blue as well?

@gcampbell-msft You're right, I did do a double take on that one. I'll revisit the approach for it! Thank you :)

File: syntaxes/CMake.tmLanguage.json (hunk 5 above)
Comment (hanniavalera) [reply syntaxes/CMake.tmLanguage.json]: Got it, will refactor!

Comment (hanniavalera) [reply syntaxes/CMake.tmLanguage.json]: Let me know your thoughts! I ended up consolidating everything into a generic rule!

**Comment by petk** [comment]
Just wanted to say, thanks so much for this. Really great!
